web/src/js/ducks/modes/utils.ts

```diff
@@ -0,0 +1,94 @@
+import { getMode as getRegularModeConfig } from "./regular";
+import { getMode as getLocalModeConfig } from "./local";
+import { fetchApi } from "../../utils";
+
+export interface ModeState {
+    active: boolean;
+    name: string;
+    listen_port?: number;
+    listen_host?: string;
+    error?: string;
+}
+
+export const updateMode = () => {
+    return async (_, getState) => {
+        try {
+            const modes = getState().modes;
+
+            const activeModes: string[] = [
+                ...getRegularModeConfig(modes),
+                ...getLocalModeConfig(modes),
+                //add new modes here
+            ];
+            const response = await fetchApi.put("/options", {
+                mode: activeModes,
+            });
+            if (response.status === 200) {
+                return { success: true };
+            } else {
+                const errorText = await response.text();
+                return { success: false, error: errorText };
+            }
+        } catch (error) {
+            //TODO: handle error
+        }
+    };
+};
+
+export const addListenAddr = (mode: ModeState) => {
+    let stringMode = mode.name;
+    if (mode.active) {
+        if (mode.listen_host && mode.listen_port) {
+            stringMode += `@${mode.listen_host}:${mode.listen_port}`;
+        } else if (mode.listen_port) {
+            stringMode += `@${mode.listen_port}`;
+        }
+        return [stringMode];
+    }
+    return [];
+};
+
+export const parseMode = (spec: string) => {
+    const [head, listenAt] = spec.includes("@") ? spec.split("@") : [spec, ""];
+
+    let mode = "";
+    let data = "";
+
+    const firstColonIndex = head.indexOf(":");
+    if (firstColonIndex !== -1) {
+        mode = head.slice(0, firstColonIndex);
+        data = head.slice(firstColonIndex + 1);
+    } else {
+        mode = head;
+    }
+
+    let host = "";
+    let port: string | number = "";
+
+    if (listenAt) {
+        if (listenAt.includes(":")) {
+            [host, port] = listenAt.split(":");
```

Comment:
If you add tests for IPv6 listen_addrs, this will fail. 😉 

We can put this into web/src/js/utils.ts:
```javascript
export function rpartition(str: string, sep: string): [string, string] {
  const lastIndex = str.lastIndexOf(sep);
  if (lastIndex === -1) {
    return ['', str];
  }
  const before = str.slice(0, lastIndex);
  const after = str.slice(lastIndex + sep.length);
  return [before, after];
}
```

Reply:
example?

Reply:
what do you think with this new version?